From dbd91734872b783443af1b62996266f67ec66598 Mon Sep 17 00:00:00 2001
From: Martin Wilck <mwilck@suse.com>
Date: Thu, 20 Jul 2017 12:40:52 +0200
Subject: string.h: add memcpy_and_pad()
Patch-mainline: Submitted to lkml, linux-nvme 2017/07/20
References: bsc#1048146

This helper function is useful for the nvme subsystem, and maybe
others.

Signed-off-by: Martin Wilck <mwilck@suse.com>
---
 include/linux/string.h |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/include/linux/string.h
+++ b/include/linux/string.h
@@ -169,4 +169,22 @@ static inline const char *kbasename(cons
 	return tail ? tail + 1 : path;
 }
 
+/**
+ * memcpy_and_pad - Copy one buffer to another with padding
+ * @dest: Where to copy to
+ * @dest_len: The destination buffer size
+ * @src: Where to copy from
+ * @count: The number of bytes to copy
+ * @pad: Character to use for padding if space is left in destination.
+ */
+static inline void memcpy_and_pad(void *dest, size_t dest_len,
+				  const void *src, size_t count, int pad)
+{
+	if (dest_len > count) {
+		memcpy(dest, src, count);
+		memset(dest + count, pad,  dest_len - count);
+	} else
+		memcpy(dest, src, dest_len);
+}
+
 #endif /* _LINUX_STRING_H_ */
